src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs

```diff
@@ -1283,7 +1283,8 @@ internal static bool IsOrContainsErrorType(TypeSymbol typeSymbol)
                         //can actually get both, so don't use else if
                         if (!hidingMemberIsNew && hiddenMember.Kind == hidingMember.Kind &&
                             !hidingMember.IsAccessor() &&
-                            (hiddenMember.IsAbstract || hiddenMember.IsVirtual || hiddenMember.IsOverride))
+                            (hiddenMember.IsAbstract || hiddenMember.IsVirtual || hiddenMember.IsOverride) &&
+                            !(hidingMember is SynthesizedRecordEquals))
```

Comment:
>hidingMember is SynthesizedRecordEquals [](start = 30, length = 39)

We should avoid type tests for particular implementation types as it makes the compiler fragile. Also, as used here, it appears to be crossing abstraction boundaries.

Please add some kind of API that can be used here instead of a type test for the intended semantic meaning.

Reply:
> We should avoid type tests for particular implementation types as it makes the compiler fragile. Also, as used here, it appears to be crossing abstraction boundaries.

> Please add some kind of API that can be used here instead of a type test for the intended semantic meaning.

In general we prefer doing things like that, and we do make exceptions in some cases where it makes sense. I think this is one of the cases like that, where addition of the new API (that we usually prefer to be abstract) for the sake of this method is not worth pollution of the API space and is not worth the work making sure every possible flavor of a Symbol implements it properly. 

---
In reply to: [450380301](https://github.com/dotnet/roslyn/pull/45647#discussion_r450380301) [](ancestors = 450380301)

Reply:
The design problem we have that causes us to want to implement such a virtual method on every symbol implementation is exactly the same problem we have with code like this, which should (for the same reason) have a case for every symbol implementation type even though only one is different. (How do you ensure every new symbol type is reviewed to see if it should be handled here?)

In cases like this encapsulating the type tests inside an extension method is a way of isolating the components. A method encapsulating a switch expression and a virtual method are complementary techniques (you can use one when design constraints prevent the other).

Reply:
>The design problem we have that causes us to want to implement such a virtual method on every symbol implementation is exactly the same problem we have with code like this, which should (for the same reason) have a case for every symbol implementation type even though only one is different. (How do you ensure every new symbol type is reviewed to see if it should be handled here?)

>In cases like this encapsulating the type tests inside an extension method is a way of isolating the components. A method encapsulating a switch expression and a virtual method are complementary techniques (you can use one when design constraints prevent the other).

Sorry, I see no difference, if we have an API we have to make sure it behaves correctly for all possible usage scenarios, where as here we only have to worry about one specific scenario. And, having an extension method (as opposed to having an abstract method) doesn't really solve the problem: "How do you ensure every new symbol type is reviewed to see if it should be handled here?" Because one has to know to look into that method, one has to know it exists. Finally, for every new symbol, the developer is responsible making sure that compiler behaves properly when the symbol is used - testing goes a long way, that is how I figured this method had to be adjusted, there was no magic arrow pointing at it.
I opened an issue [https://github.com/dotnet/roslyn/issues/45724](https://github.com/dotnet/roslyn/issues/45724) to see if we would want to do something about this later.


---
In reply to: [450433141](https://github.com/dotnet/roslyn/pull/45647#discussion_r450433141) [](ancestors = 450433141)